packages/controllers/src/cronjob/CronjobController.test.ts

```diff
@@ -0,0 +1,370 @@
+          return [getTruncatedSnap()];
+        } else if (method === 'PermissionController:getPermissions') {
+          return {
+            [SnapEndowments.Cronjob]: MOCK_TOO_FAR_CRONJOB_PERMISSION,
+          } as any;
+        }
+        return false;
+      });
+
+    const cronjobController = new CronjobController({
+      messenger: controllerMessenger,
+    });
+
+    await cronjobController.register(MOCK_SNAP_ID);
+
+    expect(callActionMock).toHaveBeenCalledWith(
+      'PermissionController:getPermissions',
+      MOCK_SNAP_ID,
+    );
+
+    // eslint-disable-next-line dot-notation
+    expect(cronjobController['_timers'].size).toBe(0);
+
+    cronjobController.destroy();
+  });
+
+  it('handles SnapInstalled event', async () => {
+    const rootMessenger = getRootCronjobControllerMessenger();
+    const controllerMessenger =
+      getRestrictedCronjobControllerMessenger(rootMessenger);
+
+    jest.spyOn(controllerMessenger, 'call').mockImplementation((method) => {
+      if (method === 'SnapController:getAll') {
+        return [getTruncatedSnap()];
+      } else if (method === 'PermissionController:getPermissions') {
+        return { [SnapEndowments.Cronjob]: MOCK_CRONJOB_PERMISSION } as any;
+      }
+      return false;
+    });
+
+    const cronjobController = new CronjobController({
+      messenger: controllerMessenger,
+    });
+
+    const snapInfo: TruncatedSnap = {
+      blocked: false,
+      enabled: true,
+      id: MOCK_SNAP_ID,
+      initialPermissions: {},
+      permissionName: '',
+      version: '',
+    };
+    // eslint-disable-next-line dot-notation
+    await cronjobController['_handleEventSnapInstalled'](snapInfo);
+
+    // eslint-disable-next-line dot-notation
+    expect(cronjobController['_snapIds'].size).toBe(2);
+
+    cronjobController.destroy();
+  });
+
+  it('handles SnapRemoved event', async () => {
+    const rootMessenger = getRootCronjobControllerMessenger();
+    const controllerMessenger =
+      getRestrictedCronjobControllerMessenger(rootMessenger);
+
+    jest.spyOn(controllerMessenger, 'call').mockImplementation((method) => {
+      if (method === 'SnapController:getAll') {
+        return [getTruncatedSnap()];
+      } else if (method === 'PermissionController:getPermissions') {
+        return { [SnapEndowments.Cronjob]: MOCK_CRONJOB_PERMISSION } as any;
+      }
+      return false;
+    });
+
+    const cronjobController = new CronjobController({
+      messenger: controllerMessenger,
+    });
+
+    await cronjobController.register(MOCK_SNAP_ID);
+
+    const snapInfo: TruncatedSnap = {
+      blocked: false,
+      enabled: true,
+      id: MOCK_SNAP_ID,
+      initialPermissions: {},
+      permissionName: '',
+      version: '',
+    };
+    // eslint-disable-next-line dot-notation
+    cronjobController['_handleEventSnapRemoved'](snapInfo);
+
+    // eslint-disable-next-line dot-notation
+    expect(cronjobController['_snapIds'].size).toBe(0);
```

Comment:
```suggestion
    cronjobController._handleEventSnapRemoved(snapInfo);

    expect(cronjobController._snapIds.size).toBe(0);
```

Reply:
The only way to make this happen `cronjobController._handleEventSnapRemoved(snapInfo);` is to make these public, but I would need some advice from @FrederikBolding since he made these private, so maybe there is a concern about it.

It happens to be very specific scenario in this case with the messenger, otherwise I would be able to test all private methods by calling the entry one that we're exposing, etc. But, in this case the tests become non-deterministic because of asynchronous ways of things with the messenger pattern that I cannot await.

For example I can publish the event like `SnapController:snapInstalled` and `_handleEventSnapInstalled` will be called then, but inside the unit test I cannot know when the registration, schedule and execution will happen exactly, so I cannot expect something to be called or to have `snapsIds`/`timers` variables updated. So, some `waiting` would have to be introduced like wait-for-expect, etc. But, since we're using `jest.useFakeTimers();` it becomes more complicated. Let me know how you've been dealing with these kind of problems in this project and I can maybe try some other ideas.

Reply:
Not sure what our pattern should be for testing private methods, we can use `ts-expect-error`...